scripts/analyze_codebase.py

```
@@ -0,0 +1,284 @@
+            sha256.update(chunk)
+    return sha256.hexdigest()
+
+
+def extract_code_blocks(filepath: Path, min_lines: int = 5) -> List[Tuple[int, int, str]]:
+    """Extract significant code blocks from a Python file.
+    
+    Returns list of (start_line, end_line, normalized_code) tuples.
+    """
+    blocks = []
+    try:
+        with open(filepath, "r", encoding="utf-8") as f:
```

Ruff (0.12.2)

Unnecessary mode argument

Remove mode argument

(UP015)